synapse/http/server.py

```diff
@@ -515,52 +515,62 @@ def __init__(
     ):
         self._request = request
         self._iterator = iterator
+        self._paused = False
 
-    def start(self) -> None:
-        self._request.registerProducer(self, False)
+        # Register the producer and start producing data.
+        self._request.registerProducer(self, True)
+        self.resumeProducing()
 
     def _send_data(self, data: List[bytes]) -> None:
         """
-        Send a list of strings as a response to the request.
+        Send a list of bytes as a chunk of a response.
         """
         if not data:
             return
         self._request.write(b"".join(data))
 
+    def pauseProducing(self) -> None:
+        self._paused = True
+
     def resumeProducing(self) -> None:
         # We've stopped producing in the meantime (note that this might be
         # re-entrant after calling write).
         if not self._request:
             return
 
-        # Get the next chunk and write it to the request.
-        #
-        # The output of the JSON encoder is coalesced until min_chunk_size is
-        # reached. (This is because JSON encoders produce a very small output
-        # per iteration.)
-        #
-        # Note that buffer stores a list of bytes (instead of appending to
-        # bytes) to hopefully avoid many allocations.
-        buffer = []
-        buffered_bytes = 0
-        while buffered_bytes < self.min_chunk_size:
-            try:
-                data = next(self._iterator)
-                buffer.append(data)
-                buffered_bytes += len(data)
-            except StopIteration:
-                # The entire JSON object has been serialized, write any
-                # remaining data, finalize the producer and the request, and
-                # clean-up any references.
-                self._send_data(buffer)
-                self._request.unregisterProducer()
-                self._request.finish()
-                self.stopProducing()
-                return
-
-        self._send_data(buffer)
+        self._paused = False
+
+        # Write until there's backpressure telling us to stop.
+        while not self._paused:
```

Comment:
This definitely looks like an infinite loop, but I don't think it is (see a previous discussion about this [description here](https://github.com/matrix-org/synapse/pull/8013#discussion_r466592298)). I think when we call `write` we give up control to the reactor which mightcall `pauseProducing` before control is returned to `resumeProducing`.

Reply:
When you call `write()` you go through the `Request` -> `HTTPChannel` -> [`iocpreactor.tcp.Connection`](https://github.com/twisted/twisted/blob/twisted-19.2.0/src/twisted/internet/iocpreactor/tcp.py#L167-L180) (this is `HTTPChannel.transport`) -> [`iocpreactor.abstract.FileHandle`](https://github.com/twisted/twisted/blob/twisted-19.2.0/src/twisted/internet/iocpreactor/abstract.py#L271-L279), which might cause `producer.pauseProducing` (which goes back through the chain in the opposite direction, essentially).

Reply:
This does mean that if we write slower than the TCP buffer drains we will write everything in one go. I don't know whether we want to put some limit to the amount written in one reactor tick? Maybe something like 64KB (or even 512KB)? A reactor tick generally takes less than 1ms, so those values would mean maxing out at ~64MB/s (or 512MB/s).

Reply:
Aha, but `FileHandle` has a write buffer of 64KB and doesn't actually pull from the buffer until a reactor tick, so we don't need to worry about it. The flow becomes: 1) write data, 2) pause producing is called if over 64kb of data has been written, 3) we exit the loop until the `FileHandle` buffer has been drained.

Reply:
That was my understanding as well! 👍 This is some fun spelunking into Twisted.